src/irmin-pack/inode.ml

```diff
@@ -660,44 +669,47 @@ struct
       | Unsorted_pointers t -> Error (`Unsorted_pointers t)
 
     let hash t = Lazy.force t.hash
+    let is_node t = t.kind = Kind.Node [@@inline]
 
     let is_root t =
       match t.v with
       | Tree { depth; _ } -> depth = 0
       | Values _ ->
-          (* When [t] is of tag [Values], then [t] is root iff [t] is stable. It
-             is implied by the following.
-
-             When [t] is stable, then [t] is a root, because:
-              - Only 2 functions produce stable inodes: [stabilize] and [empty].
-              - Only the roots are output of [stabilize].
-              - An empty map can only be located at the root.
-
-             When [t] is a root of tag [Value], then [t] is stable, because:
-             - All the roots are output of [stabilize].
-             - When an unstable inode enters [stabilize], it becomes stable if
-               it has at most [Conf.stable_hash] leaves.
-             - A [Value] has at most [Conf.stable_hash] leaves because
-               [Conf.entries <= Conf.stable_hash] is enforced.
+          (* When [t] is of tag [Values], then [t] is root iff [t] is a [Node].
+             It is implied by the following;:
+
+             When [t] is a [Node], then [t] is a root, because:
+               - Only 2 functions produce stable inodes: [stabilize] and [empty].
+               - Only the roots are output of [stabilize].
+               - An empty map can only be located at the root.
+
+             When [t] is a root of tag [Value], then [t] is a stable [Inode]
+             (hashed as a [Node]), because:
+               - All the roots are output of [stabilize].
+               - When an unstable inode enters [stabilize], it becomes a
+                 stable [Inode] if it has at most [Conf.stable_hash] leaves.
+               - A [Value] has at most [Conf.stable_hash] leaves because
+                 [Conf.entries <= Conf.stable_hash] is enforced.
           *)
-          t.stable
+          is_node t
 
     let check_write_op_supported t =
       if not @@ is_root t then
         failwith "Cannot perform operation on non-root inode value."
 
     let stabilize layout t =
-      if t.stable then t
-      else
-        let n = length t in
-        if n > Conf.stable_hash then t
-        else
-          let hash =
-            lazy
-              (let vs = list layout t in
-               Node.hash (Node.v vs))
-          in
-          { hash; stable = true; v = t.v }
+      match t.kind with
+      | Node -> t
+      | Inode ->
+          let n = length t in
+          if n > Conf.stable_hash then t
+          else
+            let hash =
+              lazy
+                (let vs = list layout t in
+                 Node.hash (Node.v vs))
+            in
+            { hash; kind = Node; v = t.v }
```

Comment:
To make it clear that we just "stabilised" the node I could replace this line with 
`{ t with hash; kind = Node }`